front/lib/api/assistant/jit_actions.ts

```diff
@@ -47,6 +57,129 @@ export async function isJITActionsEnabled(
   return use;
 }
 
+export async function listFiles(
+  auth: Authenticator,
+  { conversation }: { conversation: ConversationType }
+): Promise<ConversationFileType[]> {
+  const files: ConversationFileType[] = [];
+  for (const m of conversation.content.flat(1)) {
+    if (
+      isContentFragmentType(m) &&
+      isSupportedPlainTextContentType(m.contentType) &&
+      m.contentFragmentVersion === "latest"
+    ) {
+      if (m.fileId) {
+        files.push({
+          fileId: m.fileId,
+          title: m.title,
+          contentType: m.contentType,
+        });
+      }
+    } else if (isAgentMessageType(m)) {
+      for (const a of m.actions) {
+        if (isTablesQueryActionType(a)) {
+          if (a.resultsFileId && a.resultsFileSnippet) {
+            files.push({
+              fileId: a.resultsFileId,
+              contentType: "text/csv",
+              title: getTablesQueryResultsFileTitle({ output: a.output }),
+            });
+          }
+        }
+      }
+    }
+  }
+
+  // Fetch the snippets
+  const fileIds = files.map((f) => f.fileId);
+  const fileResources = await FileResource.fetchByIds(auth, fileIds);
+  for (const f of files) {
+    const fileResource = fileResources.find((fr) => fr.sId === f.fileId);
+    if (fileResource && fileResource.snippet) {
+      f.snippet = fileResource.snippet;
+    }
+  }
+
+  return files;
+}
+
+export async function getJITActions(
+  auth: Authenticator,
+  { conversation }: { conversation: ConversationType }
+): Promise<AgentActionConfigurationType[]> {
+  const actions: AgentActionConfigurationType[] = [];
+
+  if (await isJITActionsEnabled(auth)) {
+    const files = await listFiles(auth, { conversation });
+    if (files.length > 0) {
+      const filesUsableForJIT = files.filter((f) => !!f.snippet);
+
+      if (filesUsableForJIT.length > 0) {
+        // Get the datasource view for the conversation.
```

Comment:
cost 2 queries, not a fan, but I only have the conversationId here.

_Not sure if it's worth a dedicated function with a joined query ?_

Reply:
independent of 2 queries or one join this should totally be on DataSourceViewResource (the hint as to why is that you leverage a modelId in between the two)